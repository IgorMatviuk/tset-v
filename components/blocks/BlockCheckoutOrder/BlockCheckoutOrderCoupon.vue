<template>
  <div class="order-coupon">
    <div class="order-coupon__title">ввести промокод</div>
    <form novalidate @submit.prevent="submitCoupon">
      <div class="order-coupon__group">
        <!-- TODO: field and button animation fix -->
        <BFormInput
          ref="couponInput"
          v-model="coupon"
          :class="{ 'order-coupon__input--accepted': couponAccepted }"
          class="order-coupon__input"
          type="text"
          :readonly="couponAccepted"
          :trim="true"
          placeholder="Ввести промокод"
        />
        <transition name="fade">
          <BButton
            v-if="!couponAccepted"
            class="order-coupon__button newsletter__button"
            type="submit"
            variant="primary"
            size="sm"
          >
            <svg-icon
              class="order-coupon__icon newsletter__icon"
              name="button-arrow"
              width="27"
              height="9"
            />
          </BButton>
        </transition>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  data() {
    return {
      couponAccepted: false,
      coupon: ''
    }
  },
  methods: {
    submitCoupon() {
      // TODO: submit coupon
      if (this.coupon) {
        setTimeout(() => {
          this.couponAccepted = !this.couponAccepted
          this.$refs.couponInput.blur()
        }, 500)
      }
    }
  }
}
</script>

<style lang="scss">
.order-coupon {
  &__title {
    margin-bottom: 14px;
    font-size: 12px;
    line-height: 15px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__group {
    display: flex;
  }

  &__input {
    width: 462px;
    transition: width 0.7s;

    &--accepted {
      width: 100%;
      color: $primary;
      pointer-events: none;
    }
  }

  &__button {
    margin-left: 32px;
  }

  @include media-breakpoint-down(lg) {
    &__title {
      margin-bottom: 12px;
    }
  }

  @include media-breakpoint-down(md) {
    &__input {
      width: 572.6px;

      &--accepted {
        width: 100%;
      }
    }

    &__button {
      margin-left: 39px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__title {
      margin-bottom: 7px;
      font-size: 11px;
      line-height: 14px;
      letter-spacing: 0.05em;
    }

    &__button {
      margin-left: 19px;
    }
  }
}
</style>
