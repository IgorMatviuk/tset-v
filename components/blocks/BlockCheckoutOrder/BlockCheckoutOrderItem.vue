<template>
  <div class="order-item">
    <button class="order-item__remove" type="button" @click="removeItem">
      <svg-icon name="trash" class="order-item__icon" width="15" height="15" />
    </button>
    <div class="order-item__image-wrapper">
      <ImageLoader>
        <picture>
          <source
            :data-srcset="variant.base_image.thumb_xs | webp"
            srcset="/img/stub.png"
            type="image/webp"
          />
          <img
            :data-src="variant.base_image.thumb_xs"
            src="/img/stub.png"
            :alt="variant.name"
            :title="variant.name"
            class="order-item__image lazyload"
            loading="lazy"
            data-object-fit="cover"
          />
        </picture>
      </ImageLoader>
    </div>
    <div class="order-item__body">
      <NuxtLink :to="$i18n.path(variant.url_key)" class="order-item__name">
        {{ variant.name }}
      </NuxtLink>
      <div class="order-item__price-wrapper">
        <template v-if="variant.discount_price">
          <div class="order-item__price">
            {{ variant.formatted_discount_price }}
          </div>
          <div class="order-item__old-price">
            {{ variant.formatted_price }}
          </div>
        </template>
        <template v-else>
          <div class="order-item__price">
            {{ variant.formatted_price }}
          </div>
        </template>
      </div>
      <div class="order-item__code">
        {{ $t('productPage.sku') }}: {{ variant.sku }}
      </div>
      <!-- <OptionSelect
        :id="getOptionsKey('color', item.id)"
        :attribute-id="colorAttribute.id"
        :attribute-code="colorAttribute.code"
        :active-option="colorAttribute.selected_option"
        :options="mappedColorOptionItems"
        class="order-item__colors"
        @change="onOptionsChange"
      /> -->
      <OptionSelect
        :id="getOptionsKey('size', item.id)"
        :attribute-id="sizeAttribute.id"
        :attribute-code="sizeAttribute.code"
        :active-option="getActiveOption('size')"
        :options="mappedSizeOptionItems"
        class="order-item__sizes"
        @change="onOptionsChange"
      />
      <QuantityCounter
        :count="item.quantity"
        class="order-item__counter"
        @change="onQuantityChange"
      />
    </div>
  </div>
</template>

<script>
import OptionSelect from '@/components/shared/OptionSelect'
import QuantityCounter from '@/components/shared/QuantityCounter'
import ImageLoader from '@/components/shared/ImageLoader'

import { actionTypes as cartActionTypes } from '@/store/cart'

import productVariants from '@/mixins/product-variants'
import removeFromCart from '@/mixins/remove-from-cart'

import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    OptionSelect,
    QuantityCounter,
    ImageLoader
  },
  mixins: [productVariants, removeFromCart, imageNextGen],
  props: {
    index: {
      type: Number,
      required: true
    },
    item: {
      type: Object,
      required: true
    }
  },
  computed: {
    variant() {
      return this.item.product.variants.find(
        (variant) => variant.id === this.item.variant_id
      )
    },
    // colorAttribute() {
    //   return this.item.product.attributes.find((attr) => attr.code === 'color')
    // },
    sizeAttribute() {
      return this.item.product.attributes.find((attr) => attr.code === 'size')
    },
    // mappedColorOptionItems() {
    //   return this.mapOptionsWithStatus('color')
    // },
    mappedSizeOptionItems() {
      return this.mapOptionsWithStatus('size')
    }
  },
  methods: {
    mapOptionsWithStatus(attrCode) {
      let options
      let activeOppositeVariants

      /* istanbul ignore else */
      // if (attrCode === 'color') {
      //   options = this.colorOptions
      //   activeOppositeVariants = this.activeSizeVariants
      // }

      /* istanbul ignore else */
      if (attrCode === 'size') {
        options = this.sizeOptions
        activeOppositeVariants = this.activeColorVariants
      }

      return options.items.map((item) => {
        const intersection = item.variants.filter((variant) =>
          activeOppositeVariants.includes(variant)
        )

        return {
          disabled: !intersection.length,
          ...item
        }
      })
    },
    getOptionsKey(scope, id) {
      const capitalized = scope.charAt(0).toUpperCase() + scope.slice(1)
      return `order${capitalized}_${id}_${this.index}`
    },
    getActiveOption(attrCode) {
      const pAttrs = this.item.product.attributes
      const pVariants = this.item.product.variants

      const variant = pVariants.find((x) => x.id === this.item.variant_id)
      const vAttrs = variant.attributes

      const sourceAttr = pAttrs.find((x) => x.code === attrCode)
      const { value: optionId } = vAttrs.find((x) => x.code === attrCode)

      return sourceAttr.options.find((x) => x.id === optionId)
    },
    getIntersectedVariantId() {
      return this.activeColorVariants.find((cVariant) => {
        const sVariant = this.activeSizeVariants.includes(cVariant)
        return !!sVariant
      })
    },
    onOptionsChange({ attributeCode, attributeId, option }) {
      /* istanbul ignore else */
      // if (attributeCode === 'color') {
      //   this.activeColorVariants = option.variants
      // }

      /* istanbul ignore else */
      if (attributeCode === 'size') {
        this.activeSizeVariants = option.variants
      }

      const variantId = this.getIntersectedVariantId()

      this.updateItem(variantId, this.item.quantity)
    },
    onQuantityChange(value) {
      this.updateItem(null, value)
    },
    updateItem(variant, quantity) {
      const payload = {
        cartId: this.$store.state.localStorage.cartId,
        itemId: this.item.id,
        body: { quantity }
      }

      /* istanbul ignore else */
      if (variant) {
        payload.body.variant = variant
      }

      this.$store.dispatch(`cart/${cartActionTypes.UPDATE_ITEM}`, payload)
    }
  }
}
</script>

<style lang="scss">
.order-item {
  display: flex;
  align-items: flex-start;

  &__remove {
    @include reset-button-styles();

    display: flex;
    flex: 0 0 30px;
    width: 30px;
    height: 30px;
    margin-top: -8px;
    transition: color 0.3s;
  }

  &__remove:hover &__icon,
  &__remove:focus &__icon,
  &__remove:active &__icon {
    color: $primary;
  }

  &__icon {
    flex: 0 0 15px;
    width: 15px;
    height: 15px;
    color: $text-color;
  }

  &__image-wrapper {
    position: relative;
    flex: 0 0 70px;
    height: 90px;
    margin-right: 14px;
  }

  &__image {
    max-width: 100%;
    max-height: 100%;
    object-fit: cover;
  }

  &__body {
    display: flex;
    flex-grow: 1;
    flex-wrap: wrap;
  }

  &__name {
    flex: 0 0 275px;
    max-width: 275px;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    font-size: 12px;
    line-height: 15px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__price-wrapper {
    margin-left: auto;
    text-align: right;
  }

  &__old-price {
    font-size: 11px;
    line-height: 14px;
    text-decoration-line: line-through;
    color: $text-color;
  }

  &__price {
    font-size: 14px;
    font-weight: 500;
    line-height: 18px;
    color: $primary;
  }

  &__code {
    width: 100%;
    margin-bottom: 17px;
    font-size: 12px;
    line-height: 15px;
    color: $gray-dark-1;
  }

  &__colors,
  &__sizes {
    align-self: center;
    margin-bottom: 2px;
  }

  &__colors {
    margin-right: 23px;
  }

  &__counter {
    margin-left: auto;
  }

  @include media-breakpoint-down(xl) {
    &__colors,
    &__sizes {
      margin-bottom: 4px;
    }
  }

  @include media-breakpoint-down(lg) {
    &__remove {
      justify-content: center;
      padding-right: 5px;
      margin-top: -9px;
    }

    &__icon {
      width: 13px;
      height: 13px;
    }

    &__image-wrapper {
      margin-right: 10px;
    }

    &__body {
      margin-top: -3px;
    }

    &__name {
      display: -webkit-box;
      flex: 0 0 170px;
      max-width: 170px;
    }

    &__code {
      margin-bottom: 19px;
      font-size: 11px;
      line-height: 17px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 0;
    }

    &__colors {
      margin-right: 15px;
    }
  }

  @include media-breakpoint-down(md) {
    &__remove {
      justify-content: flex-start;
    }

    &__image-wrapper {
      margin-right: 15px;
    }

    &__body {
      margin-top: 0;
    }

    &__name {
      flex: 0 0 275px;
      max-width: 275px;
    }

    &__code {
      margin-bottom: 15px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 4px;
    }

    &__colors {
      margin-right: 25px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__remove {
      flex: 0 0 23px;
      width: 23px;
      padding-right: 0;
    }

    &__image-wrapper {
      margin-right: 12px;
    }

    &__body {
      margin-top: -3px;
    }

    &__name {
      flex: 0 0 125px;
      max-width: 125px;
    }

    &__code {
      margin-bottom: 23px;
    }

    &__colors,
    &__sizes {
      margin-bottom: 1px;
    }

    &__colors {
      margin-right: 10px;
    }
  }
}
</style>
