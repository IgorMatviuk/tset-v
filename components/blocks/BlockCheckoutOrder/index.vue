<template>
  <div class="checkout-order">
    <div class="checkout-order__header">
      <div class="checkout-order__title">
        {{ $t('checkoutPage.yourOrder') }}
      </div>
      <NuxtLink :to="backToShoppingLink" class="checkout-order__proceed">
        {{ $t('checkoutPage.backToShopping') }}
      </NuxtLink>
    </div>
    <div class="checkout-order__content">
      <BlockCheckoutOrderItem
        v-for="(item, index) in orderSummary.items"
        :key="item.variant_id"
        :page="'checkout'"
        :item="item"
        :index="index"
        class="checkout-order__item"
      />
    </div>
    <!-- TODO: uncomment for the future release -->
    <!-- <BlockCheckoutOrderCoupon class="checkout-order__coupon" /> -->
    <div class="checkout-order__footer">
      <transition-group name="fade">
        <div
          v-if="orderSummary.discount_price"
          key="price"
          class="checkout-order__footer-item checkout-order__footer-item--sale"
        >
          <div
            class="checkout-order__footer-title checkout-order__footer-title--sale"
          >
            {{ $t('checkoutPage.orderPrice') }}
          </div>
          <div class="checkout-order__price">
            {{ orderSummary.formatted_grand_total }}
          </div>
        </div>
        <div
          v-if="orderSummary.discount_price"
          key="sale"
          class="checkout-order__footer-item"
        >
          <div
            class="checkout-order__footer-title checkout-order__footer-title--sale"
          >
            {{ $t('checkoutPage.discount') }}
          </div>
          <div class="checkout-order__sale">
            - {{ orderSummary.grand_total - orderSummary.discount_price }} грн
          </div>
        </div>
        <div
          v-if="orderSummary.grand_total > 1000"
          key="freeShipping"
          class="checkout-order__footer-item"
        >
          <div class="checkout-order__footer-title">
            {{ $t('checkoutPage.freeShipping.name') }}
          </div>
          <div class="checkout-order__total-price">
            {{ $t('checkoutPage.freeShipping.price') }}
          </div>
        </div>
        <div key="totalPrice" class="checkout-order__footer-item">
          <div class="checkout-order__footer-title">
            {{ $t('cart.total') }}
          </div>
          <div class="checkout-order__total-price">
            {{
              orderSummary.discount_price
                ? orderSummary.formatted_discount_price
                : orderSummary.formatted_grand_total
            }}
          </div>
        </div>
      </transition-group>
    </div>
  </div>
</template>

<script>
import BlockCheckoutOrderItem from './BlockCheckoutOrderItem'
// import BlockCheckoutOrderCoupon from './BlockCheckoutOrderCoupon'

export default {
  components: {
    BlockCheckoutOrderItem
    // BlockCheckoutOrderCoupon
  },
  props: {
    orderSummary: {
      type: Object,
      required: true
    }
  },
  computed: {
    backToShoppingLink() {
      return this.$i18n.path(this.$store.getters['catalog/defaultCategorySlug'])
    }
  }
}
</script>

<style lang="scss">
.checkout-order {
  &__header,
  &__footer-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  &__title {
    font-size: 14px;
    line-height: 18px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__proceed {
    font-size: 12px;
    font-weight: 300;
    line-height: 15px;
    letter-spacing: 0.05em;
    text-decoration-line: underline;

    @include hover-focus-active {
      text-decoration-line: underline;
    }
  }

  &__content {
    padding-top: 19px;
    margin-bottom: 23px;
  }

  &__item {
    position: relative;

    &:not(:last-of-type) {
      margin-bottom: 51px;

      &::after {
        position: absolute;
        right: 0;
        bottom: -26px;
        left: 28px;
        height: 1px;
        content: '';
        background-color: $gray;
      }
    }
  }

  &__coupon {
    padding-top: 15px;
    margin-bottom: 20px;
    margin-left: 27px;
    border-top: 1px solid $gray;
  }

  &__footer {
    margin-left: 27px;
  }

  &__footer-item {
    padding-bottom: 13px;
    border-bottom: 1px solid $gray;

    &:not(:last-of-type) {
      margin-bottom: 15px;
    }

    &--sale {
      padding-top: 12px;
    }
  }

  &__footer-title {
    font-size: 12px;
    line-height: 15px;
    text-transform: uppercase;
    letter-spacing: 0.05em;

    &--sale {
      text-transform: none;
    }
  }

  &__price,
  &__sale {
    font-size: 16px;
    line-height: 19px;
    color: $input-color;
  }

  &__sale {
    color: $primary;
  }

  &__total-price {
    font-size: 18px;
    font-weight: 500;
    line-height: 23px;
    color: $input-color;
  }

  @include media-breakpoint-down(lg) {
    &__header {
      padding-left: 6px;
    }

    &__content {
      margin-bottom: 16px;
    }

    &__item {
      &:not(:last-of-type) {
        margin-bottom: 34px;

        &::after {
          bottom: -17px;
        }
      }
    }

    &__coupon {
      padding-top: 30px;
    }
  }

  @include media-breakpoint-down(md) {
    &__header {
      padding-left: 0;
    }

    &__content {
      margin-bottom: 20px;
    }

    &__item {
      &:not(:last-of-type) {
        margin-bottom: 40px;

        &::after {
          bottom: -20px;
        }
      }
    }
  }

  @include media-breakpoint-down(sm) {
    &__content {
      padding-top: 27px;
      margin-bottom: 19px;
    }

    &__item {
      &:not(:last-of-type) {
        &::after {
          left: 0;
        }
      }
    }

    &__coupon {
      padding-top: 14px;
      margin-bottom: 24px;
      margin-left: 0;
    }

    &__footer {
      margin-left: 0;
    }

    &__footer-item {
      padding-bottom: 10px;
    }

    &__footer-title {
      font-size: 11px;
      line-height: 14px;
      letter-spacing: 0.05em;
    }

    &__price {
      font-size: 11px;
      line-height: 14px;
    }

    &__sale {
      font-size: 11px;
      line-height: 14px;
    }

    &__total-price {
      font-size: 13px;
      line-height: 16px;
    }
  }
}
</style>
