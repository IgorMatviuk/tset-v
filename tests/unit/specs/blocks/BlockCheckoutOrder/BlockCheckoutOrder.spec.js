import { shallowMount } from '@vue/test-utils'

import BlockCheckoutOrder from '@/components/blocks/BlockCheckoutOrder'

describe('BlockCheckoutOrder', () => {
  function createCheckoutOrder(orderSummary = {}) {
    const wrapper = shallowMount(BlockCheckoutOrder, {
      propsData: { orderSummary },
      mocks: {
        $t: (str) => str,
        $i18n: {
          path: (str) => str
        },
        $store: {
          getters: {}
        }
      },
      stubs: ['NuxtLink']
    })

    return wrapper
  }

  beforeEach(() => {})

  test('renders discount prices', () => {
    const wrapper = createCheckoutOrder({
      discount_price: '800.0000',
      formatted_discount_price: '800 грн',
      grand_total: 1000
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders free shipping message', () => {
    const wrapper = createCheckoutOrder({
      discount_price: '1600.0000',
      formatted_discount_price: '1600 грн',
      grand_total: 2000
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
