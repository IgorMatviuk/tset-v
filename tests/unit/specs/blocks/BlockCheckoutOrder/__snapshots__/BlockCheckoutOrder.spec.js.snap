// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockCheckoutOrder renders discount prices 1`] = `
<div
  class="checkout-order"
>
  <div
    class="checkout-order__header"
  >
    <div
      class="checkout-order__title"
    >
      
      checkoutPage.yourOrder
    
    </div>
     
    <nuxtlink-stub
      class="checkout-order__proceed"
    >
      
      checkoutPage.backToShopping
    
    </nuxtlink-stub>
  </div>
   
  <div
    class="checkout-order__content"
  />
   
  <div
    class="checkout-order__footer"
  >
    <transition-group-stub
      name="fade"
    >
      <div
        class="checkout-order__footer-item checkout-order__footer-item--sale"
      >
        <div
          class="checkout-order__footer-title checkout-order__footer-title--sale"
        >
          
          checkoutPage.orderPrice
        
        </div>
         
        <div
          class="checkout-order__price"
        >
          
          
        
        </div>
      </div>
       
      <div
        class="checkout-order__footer-item"
      >
        <div
          class="checkout-order__footer-title checkout-order__footer-title--sale"
        >
          
          checkoutPage.discount
        
        </div>
         
        <div
          class="checkout-order__sale"
        >
          
          - 200 грн
        
        </div>
      </div>
       
      <!---->
       
      <div
        class="checkout-order__footer-item"
      >
        <div
          class="checkout-order__footer-title"
        >
          
          cart.total
        
        </div>
         
        <div
          class="checkout-order__total-price"
        >
          
          800 грн
        
        </div>
      </div>
    </transition-group-stub>
  </div>
</div>
`;

exports[`BlockCheckoutOrder renders free shipping message 1`] = `
<div
  class="checkout-order"
>
  <div
    class="checkout-order__header"
  >
    <div
      class="checkout-order__title"
    >
      
      checkoutPage.yourOrder
    
    </div>
     
    <nuxtlink-stub
      class="checkout-order__proceed"
    >
      
      checkoutPage.backToShopping
    
    </nuxtlink-stub>
  </div>
   
  <div
    class="checkout-order__content"
  />
   
  <div
    class="checkout-order__footer"
  >
    <transition-group-stub
      name="fade"
    >
      <div
        class="checkout-order__footer-item checkout-order__footer-item--sale"
      >
        <div
          class="checkout-order__footer-title checkout-order__footer-title--sale"
        >
          
          checkoutPage.orderPrice
        
        </div>
         
        <div
          class="checkout-order__price"
        >
          
          
        
        </div>
      </div>
       
      <div
        class="checkout-order__footer-item"
      >
        <div
          class="checkout-order__footer-title checkout-order__footer-title--sale"
        >
          
          checkoutPage.discount
        
        </div>
         
        <div
          class="checkout-order__sale"
        >
          
          - 400 грн
        
        </div>
      </div>
       
      <div
        class="checkout-order__footer-item"
      >
        <div
          class="checkout-order__footer-title"
        >
          
          checkoutPage.freeShipping.name
        
        </div>
         
        <div
          class="checkout-order__total-price"
        >
          
          checkoutPage.freeShipping.price
        
        </div>
      </div>
       
      <div
        class="checkout-order__footer-item"
      >
        <div
          class="checkout-order__footer-title"
        >
          
          cart.total
        
        </div>
         
        <div
          class="checkout-order__total-price"
        >
          
          1600 грн
        
        </div>
      </div>
    </transition-group-stub>
  </div>
</div>
`;
